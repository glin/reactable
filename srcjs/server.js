import React from 'react'
import * as ReactDOMServer from 'react-dom/server'
import TestRenderer from 'react-test-renderer'
import createEmotionServer from '@emotion/server/create-instance'

import { Reactable } from './Reactable'
import { ReactableData } from './Reactable'
import { evaluateStringMember } from './htmlwidgets'
import './intl-polyfill'
import { getEmotion } from './theme'

// Provide React as a global for users that call React externally in custom JS render functions.
// Although `global` may not necessarily be available in V8, it is in reactable's V8 context,
// and Webpack transforms this to a portable assignment that doesn't assume `global` exists anyway.
global.React = React

const cache = getEmotion().cache
const { extractCritical } = createEmotionServer(cache)

// Render to HTML for static rendering
export function renderToHTML(inputJson) {
  const input = JSON.parse(inputJson)

  const props = input.props

  // Resolve strings marked as JavaScript literals to objects
  if (input.evals) {
    for (let key of input.evals) {
      evaluateStringMember(props, key)
    }
  }

  const { html, css, ids } = extractCritical(
    ReactDOMServer.renderToString(<Reactable {...props} />)
  )
  return { html, css, ids }
}

let initialProps
export function setInitialProps(inputJson) {
  const start = Date.now()
  const input = JSON.parse(inputJson)
  const { props, evals } = input
  // Resolve strings marked as JavaScript literals to objects
  if (evals) {
    for (let key of evals) {
      evaluateStringMember(props, key)
    }
  }
  initialProps = props
  const end = Date.now()
  debugLog('setInitialProps() JSON parse time: ', end - start)
  debugLog('setInitialProps() initial props: ', JSON.stringify(initialProps, null, 2))
}

let testRenderer

// Render to JSON data for server-side data processing
export function renderToData(inputJson) {
  const input = JSON.parse(inputJson)
  debugLog('renderToData() input: ', JSON.stringify(input, null, 2))
  const { props, evals } = input

  // Resolve strings marked as JavaScript literals to objects, e.g., searchMethod
  if (evals) {
    for (let key of evals) {
      evaluateStringMember(props, key)
    }
  }

  let result
  const setResolvedData = data => {
    result = data
  }

  if (!testRenderer) {
    const start = Date.now()
    testRenderer = TestRenderer.create(
      <ReactableData {...initialProps} {...props} setResolvedData={setResolvedData} />
    )
    const end = Date.now()
    debugLog('TestRenderer create time: ', end - start)
  } else {
    const start = Date.now()
    testRenderer.update(
      <ReactableData {...initialProps} {...props} setResolvedData={setResolvedData} />
    )
    const end = Date.now()
    debugLog('TestRenderer update time: ', end - start)
  }

  return result
}

// Only used for testing
export function resetTestRenderer() {
  initialProps = null
  testRenderer = null
}

let debugLoggingEnabled
export function enableDebugLogging(enabled = true) {
  debugLoggingEnabled = enabled
}

function debugLog(...args) {
  if (!debugLoggingEnabled) {
    return
  }
  console.log('DEBUG ', ...args)
}
