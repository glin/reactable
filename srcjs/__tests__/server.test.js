/**
 * @jest-environment node
 */

import { renderToHTML, setInitialProps, renderToData, resetTestRenderer } from '../server'

// These tests run in Node and are intentionally minimal. See test-reactable.R for
// more comprehensive tests that run in true V8 environments (different from Node).

describe('renderToHTML', () => {
  it('renders tables to HTML', () => {
    const inputString =
      '{"props":{"data":{"Manufacturer":["Acura","Acura","Audi"],"Model":["Integra","Legend","90"],"Type":["Small","Midsize","Compact"]},"columns":[{"id":"Manufacturer","name":"Manufacturer","type":"factor"},{"id":"Model","name":"Model","type":"factor"},{"id":"Type","name":"Type","type":"factor"}],"dataKey":"b540d0d929ae2ade6248d0cbb7f477fc"},"evals":[]}'
    const { html, css, ids } = renderToHTML(inputString)
    expect(html).toMatchSnapshot()
    expect(css).toEqual('')
    expect(ids).toEqual([])
  })

  it('renders tables to HTML with JS evals', () => {
    const inputString =
      '{"props":{"data":{"Manufacturer":["Acura","Acura"],"Model":["Integra","Legend"]},"columns":[{"id":"Manufacturer","name":"Manufacturer","type":"factor"},{"id":"Model","name":"Model","type":"factor","cell":"cellInfo => cellInfo.value + \'__\'"}],"dataKey":"7951e598fecb5fa57c37a854a0917a7c"},"evals":["columns.1.cell"]}'
    const { html, css, ids } = renderToHTML(inputString)
    expect(html).toContain('Integra__')
    expect(html).toContain('Legend__')
    expect(html).toMatchSnapshot()
    expect(css).toEqual('')
    expect(ids).toEqual([])
  })

  it('renders tables to HTML with theme styles', () => {
    const inputString =
      '{"props":{"data":{"Manufacturer":["Acura","Acura"],"Model":["Integra","Legend"]},"columns":[{"id":"Manufacturer","name":"Manufacturer","type":"factor"},{"id":"Model","name":"Model","type":"factor"}],"theme":{"color":"red","cellPadding":"1rem"},"dataKey":"01768d44dd8153352ff1cfa908f38d72"},"evals":[]}'
    const { html, css, ids } = renderToHTML(inputString)
    expect(html).toMatchSnapshot()
    expect(css).toEqual('.reactable-1galy0v{color:red;}.reactable-1rsvkai{padding:1rem;}')
    expect(ids).toEqual(['1galy0v', '1xdhyk6', '1rsvkai'])
  })
})

describe('renderToData', () => {
  beforeEach(() => {
    resetTestRenderer()
  })

  it('renders tables to data', () => {
    const propsJson = JSON.stringify({
      props: {
        data: {
          Manufacturer: ['Acura', 'Acura'],
          Model: ['Integra', 'Legend'],
          Type: ['Small', 'Midsize']
        },
        columns: [
          { id: 'Manufacturer', name: 'Manufacturer', type: 'factor' },
          { id: 'Model', name: 'Model', type: 'factor' },
          { id: 'Type', name: 'Type', type: 'factor' }
        ],
        pagination: true,
        paginateSubRows: false
      }
    })
    setInitialProps(propsJson)

    const inputJson = `{
      "props": {
        "pageIndex": 0,
        "pageSize": 10
      }
    }`

    const data = renderToData(inputJson)
    const expected = {
      data: [
        { Manufacturer: 'Acura', Model: 'Integra', Type: 'Small', __state: { id: '0', index: 0 } },
        { Manufacturer: 'Acura', Model: 'Legend', Type: 'Midsize', __state: { id: '1', index: 1 } }
      ],
      rowCount: 2,
      maxRowCount: 2
    }
    expect(data).toEqual(expected)
  })

  it('renders tables to data with JS evals', () => {
    const propsJson = JSON.stringify({
      props: {
        data: {
          Manufacturer: ['Acura', 'Acura'],
          Model: ['Integra', 'Legend'],
          Type: ['Small', 'Midsize']
        },
        columns: [
          { id: 'Manufacturer', name: 'Manufacturer', type: 'factor' },
          { id: 'Model', name: 'Model', type: 'factor', aggregate: "() => '__aggregated__'" },
          { id: 'Type', name: 'Type', type: 'factor' }
        ],
        pagination: true,
        paginateSubRows: false
      },
      evals: ['columns.1.aggregate']
    })
    setInitialProps(propsJson)

    const inputJson = JSON.stringify({
      props: { pageIndex: 0, pageSize: 10, groupBy: ['Manufacturer'] }
    })
    const data = renderToData(inputJson)
    expect(data.data[0].Model).toEqual('__aggregated__')
    expect(data).toMatchSnapshot()
  })

  it('renders tables to data with updates', () => {
    const propsJson = JSON.stringify({
      props: {
        data: {
          Manufacturer: ['Acura', 'Acura', 'Audi', 'Audi', 'BMW'],
          Model: ['Integra', 'Legend', '90', '100', '535i'],
          Type: ['Small', 'Midsize', 'Compact', 'Midsize', 'Midsize']
        },
        columns: [
          { id: 'Manufacturer', name: 'Manufacturer', type: 'factor' },
          { id: 'Model', name: 'Model', type: 'factor' },
          { id: 'Type', name: 'Type', type: 'factor' }
        ],
        pagination: true,
        paginateSubRows: false
      }
    })
    setInitialProps(propsJson)

    let inputJson = JSON.stringify({
      props: {
        pageIndex: 0,
        pageSize: 3,
        sortBy: [{ id: 'Manufacturer', desc: false }]
      }
    })
    let data = renderToData(inputJson)
    expect(data).toMatchSnapshot()

    inputJson = JSON.stringify({
      props: {
        pageIndex: 1,
        pageSize: 3,
        sortBy: [{ id: 'Manufacturer', desc: false }]
      }
    })
    data = renderToData(inputJson)
    expect(data).toMatchSnapshot()
  })

  it('sets maxRowCount correctly when data is filtered', () => {
    const propsJson = JSON.stringify({
      props: {
        data: {
          Manufacturer: ['Acura', 'Acura', 'Audi', 'Audi', 'BMW'],
          Model: ['Integra', 'Legend', '90', '100', '535i'],
          Type: ['Small', 'Midsize', 'Compact', 'Midsize', 'Midsize']
        },
        columns: [
          { id: 'Manufacturer', name: 'Manufacturer', type: 'factor' },
          { id: 'Model', name: 'Model', type: 'factor' },
          { id: 'Type', name: 'Type', type: 'factor' }
        ],
        pagination: true,
        paginateSubRows: false
      }
    })
    setInitialProps(propsJson)

    // Initial render without filtering to initialize maxRowCount
    let inputJson = JSON.stringify({
      props: {
        pageIndex: 0,
        pageSize: 10
      }
    })
    let data = renderToData(inputJson)
    expect(data.rowCount).toEqual(5)
    expect(data.maxRowCount).toEqual(5)
    expect(data).toMatchSnapshot()

    inputJson = JSON.stringify({
      props: {
        pageIndex: 0,
        pageSize: 10,
        searchValue: 'midsize'
      }
    })
    data = renderToData(inputJson)
    expect(data.rowCount).toEqual(3)
    expect(data.maxRowCount).toEqual(5)
    expect(data).toMatchSnapshot()
  })

  it('sets maxRowCount correctly when data is grouped, paginateSubRows=true', () => {
    const propsJson = JSON.stringify({
      props: {
        data: {
          Manufacturer: ['Acura', 'Acura', 'Audi', 'Audi', 'BMW', 'Buick', 'Buick'],
          Model: ['Integra', 'Legend', '90', '100', '535i', 'Century', 'LeSabre'],
          Type: ['Small', 'Midsize', 'Compact', 'Midsize', 'Midsize', 'Midsize', 'Large'],
          'Min.Price': [12.9, 29.2, 25.9, 30.8, 23.7, 14.2, 19.9],
          Price: [15.9, 33.9, 29.1, 37.7, 30, 15.7, 20.8]
        },
        columns: [
          { id: 'Manufacturer', name: 'Manufacturer', type: 'factor' },
          { id: 'Model', name: 'Model', type: 'factor' },
          { id: 'Type', name: 'Type', type: 'factor' },
          { id: 'Min.Price', name: 'Min.Price', type: 'numeric' },
          { id: 'Price', name: 'Price', type: 'numeric' }
        ],
        pagination: true,
        paginateSubRows: true
      }
    })
    setInitialProps(propsJson)

    const inputJson = JSON.stringify({
      props: { pageIndex: 0, pageSize: 10, groupBy: ['Manufacturer'] }
    })
    const data = renderToData(inputJson)
    expect(data.rowCount).toEqual(4)
    expect(data.maxRowCount).toEqual(11)
    expect(data).toMatchSnapshot()
  })
})
