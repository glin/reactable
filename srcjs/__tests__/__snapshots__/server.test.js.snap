// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renderToData renders tables to data with JS evals 1`] = `
{
  "data": [
    {
      ".subRows": [
        {
          "Manufacturer": "Acura",
          "Model": "Integra",
          "Type": "Small",
          "__state": {
            "id": "0",
            "index": 0,
          },
        },
        {
          "Manufacturer": "Acura",
          "Model": "Legend",
          "Type": "Midsize",
          "__state": {
            "id": "1",
            "index": 1,
          },
        },
      ],
      "Manufacturer": "Acura",
      "Model": "__aggregated__",
      "Type": null,
      "__state": {
        "grouped": true,
        "id": "Manufacturer:Acura",
      },
    },
  ],
  "maxRowCount": 1,
  "rowCount": 1,
}
`;

exports[`renderToData renders tables to data with updates 1`] = `
{
  "data": [
    {
      "Manufacturer": "Acura",
      "Model": "Integra",
      "Type": "Small",
      "__state": {
        "id": "0",
        "index": 0,
      },
    },
    {
      "Manufacturer": "Acura",
      "Model": "Legend",
      "Type": "Midsize",
      "__state": {
        "id": "1",
        "index": 1,
      },
    },
    {
      "Manufacturer": "Audi",
      "Model": "90",
      "Type": "Compact",
      "__state": {
        "id": "2",
        "index": 2,
      },
    },
  ],
  "maxRowCount": 5,
  "rowCount": 5,
}
`;

exports[`renderToData renders tables to data with updates 2`] = `
{
  "data": [
    {
      "Manufacturer": "Audi",
      "Model": "100",
      "Type": "Midsize",
      "__state": {
        "id": "3",
        "index": 3,
      },
    },
    {
      "Manufacturer": "BMW",
      "Model": "535i",
      "Type": "Midsize",
      "__state": {
        "id": "4",
        "index": 4,
      },
    },
  ],
  "maxRowCount": 5,
  "rowCount": 5,
}
`;

exports[`renderToData sets maxRowCount correctly when data is filtered 1`] = `
{
  "data": [
    {
      "Manufacturer": "Acura",
      "Model": "Integra",
      "Type": "Small",
      "__state": {
        "id": "0",
        "index": 0,
      },
    },
    {
      "Manufacturer": "Acura",
      "Model": "Legend",
      "Type": "Midsize",
      "__state": {
        "id": "1",
        "index": 1,
      },
    },
    {
      "Manufacturer": "Audi",
      "Model": "90",
      "Type": "Compact",
      "__state": {
        "id": "2",
        "index": 2,
      },
    },
    {
      "Manufacturer": "Audi",
      "Model": "100",
      "Type": "Midsize",
      "__state": {
        "id": "3",
        "index": 3,
      },
    },
    {
      "Manufacturer": "BMW",
      "Model": "535i",
      "Type": "Midsize",
      "__state": {
        "id": "4",
        "index": 4,
      },
    },
  ],
  "maxRowCount": 5,
  "rowCount": 5,
}
`;

exports[`renderToData sets maxRowCount correctly when data is filtered 2`] = `
{
  "data": [
    {
      "Manufacturer": "Acura",
      "Model": "Legend",
      "Type": "Midsize",
      "__state": {
        "id": "1",
        "index": 1,
      },
    },
    {
      "Manufacturer": "Audi",
      "Model": "100",
      "Type": "Midsize",
      "__state": {
        "id": "3",
        "index": 3,
      },
    },
    {
      "Manufacturer": "BMW",
      "Model": "535i",
      "Type": "Midsize",
      "__state": {
        "id": "4",
        "index": 4,
      },
    },
  ],
  "maxRowCount": 5,
  "rowCount": 3,
}
`;

exports[`renderToData sets maxRowCount correctly when data is grouped, paginateSubRows=true 1`] = `
{
  "data": [
    {
      "Manufacturer": "Acura",
      "Min.Price": null,
      "Model": null,
      "Price": null,
      "Type": null,
      "__state": {
        "grouped": true,
        "id": "Manufacturer:Acura",
        "subRowCount": 2,
      },
    },
    {
      "Manufacturer": "Audi",
      "Min.Price": null,
      "Model": null,
      "Price": null,
      "Type": null,
      "__state": {
        "grouped": true,
        "id": "Manufacturer:Audi",
        "subRowCount": 2,
      },
    },
    {
      "Manufacturer": "BMW",
      "Min.Price": null,
      "Model": null,
      "Price": null,
      "Type": null,
      "__state": {
        "grouped": true,
        "id": "Manufacturer:BMW",
        "subRowCount": 1,
      },
    },
    {
      "Manufacturer": "Buick",
      "Min.Price": null,
      "Model": null,
      "Price": null,
      "Type": null,
      "__state": {
        "grouped": true,
        "id": "Manufacturer:Buick",
        "subRowCount": 2,
      },
    },
  ],
  "maxRowCount": 11,
  "rowCount": 4,
}
`;

exports[`renderToHTML renders tables to HTML 1`] = `"<div class="Reactable ReactTable"><div class="rt-table" role="table"><div class="rt-thead" role="rowgroup" style="min-width:300px"><div class="rt-tr rt-tr-header" role="row" style="flex:1 0 auto;min-width:300px"><div class="rt-th rt-align-left" role="columnheader" style="position:relative;flex:100 0 auto;min-width:100px;width:100px" aria-label="Sort Manufacturer" aria-sort="none" tabindex="0" data-sort-hint="ascending"><div class="rt-th-inner"><div class="rt-sort-header"><div class="rt-text-content">Manufacturer</div><span class="rt-sort-right" aria-hidden="true"></span></div></div></div><div class="rt-th rt-align-left" role="columnheader" style="position:relative;flex:100 0 auto;min-width:100px;width:100px" aria-label="Sort Model" aria-sort="none" tabindex="0" data-sort-hint="ascending"><div class="rt-th-inner"><div class="rt-sort-header"><div class="rt-text-content">Model</div><span class="rt-sort-right" aria-hidden="true"></span></div></div></div><div class="rt-th rt-align-left" role="columnheader" style="position:relative;flex:100 0 auto;min-width:100px;width:100px" aria-label="Sort Type" aria-sort="none" tabindex="0" data-sort-hint="ascending"><div class="rt-th-inner"><div class="rt-sort-header"><div class="rt-text-content">Type</div><span class="rt-sort-right" aria-hidden="true"></span></div></div></div></div></div><div class="rt-tbody" role="rowgroup" style="min-width:300px"><div class="rt-tr-group"><div class="rt-tr" role="row" style="flex:1 0 auto;min-width:300px"><div class="rt-td rt-align-left" role="cell" style="flex:100 0 auto;min-width:100px;width:100px"><div class="rt-td-inner">Acura</div></div><div class="rt-td rt-align-left" role="cell" style="flex:100 0 auto;min-width:100px;width:100px"><div class="rt-td-inner">Integra</div></div><div class="rt-td rt-align-left" role="cell" style="flex:100 0 auto;min-width:100px;width:100px"><div class="rt-td-inner">Small</div></div></div></div><div class="rt-tr-group"><div class="rt-tr" role="row" style="flex:1 0 auto;min-width:300px"><div class="rt-td rt-align-left" role="cell" style="flex:100 0 auto;min-width:100px;width:100px"><div class="rt-td-inner">Acura</div></div><div class="rt-td rt-align-left" role="cell" style="flex:100 0 auto;min-width:100px;width:100px"><div class="rt-td-inner">Legend</div></div><div class="rt-td rt-align-left" role="cell" style="flex:100 0 auto;min-width:100px;width:100px"><div class="rt-td-inner">Midsize</div></div></div></div><div class="rt-tr-group"><div class="rt-tr" role="row" style="flex:1 0 auto;min-width:300px"><div class="rt-td rt-align-left" role="cell" style="flex:100 0 auto;min-width:100px;width:100px"><div class="rt-td-inner">Audi</div></div><div class="rt-td rt-align-left" role="cell" style="flex:100 0 auto;min-width:100px;width:100px"><div class="rt-td-inner">90</div></div><div class="rt-td rt-align-left" role="cell" style="flex:100 0 auto;min-width:100px;width:100px"><div class="rt-td-inner">Compact</div></div></div></div><div class="rt-no-data" aria-live="assertive"></div></div></div></div>"`;

exports[`renderToHTML renders tables to HTML with JS evals 1`] = `"<div class="Reactable ReactTable"><div class="rt-table" role="table"><div class="rt-thead" role="rowgroup" style="min-width:200px"><div class="rt-tr rt-tr-header" role="row" style="flex:1 0 auto;min-width:200px"><div class="rt-th rt-align-left" role="columnheader" style="position:relative;flex:100 0 auto;min-width:100px;width:100px" aria-label="Sort Manufacturer" aria-sort="none" tabindex="0" data-sort-hint="ascending"><div class="rt-th-inner"><div class="rt-sort-header"><div class="rt-text-content">Manufacturer</div><span class="rt-sort-right" aria-hidden="true"></span></div></div></div><div class="rt-th rt-align-left" role="columnheader" style="position:relative;flex:100 0 auto;min-width:100px;width:100px" aria-label="Sort Model" aria-sort="none" tabindex="0" data-sort-hint="ascending"><div class="rt-th-inner"><div class="rt-sort-header"><div class="rt-text-content">Model</div><span class="rt-sort-right" aria-hidden="true"></span></div></div></div></div></div><div class="rt-tbody" role="rowgroup" style="min-width:200px"><div class="rt-tr-group"><div class="rt-tr" role="row" style="flex:1 0 auto;min-width:200px"><div class="rt-td rt-align-left" role="cell" style="flex:100 0 auto;min-width:100px;width:100px"><div class="rt-td-inner">Acura</div></div><div class="rt-td rt-align-left" role="cell" style="flex:100 0 auto;min-width:100px;width:100px"><div class="rt-td-inner">Integra__</div></div></div></div><div class="rt-tr-group"><div class="rt-tr" role="row" style="flex:1 0 auto;min-width:200px"><div class="rt-td rt-align-left" role="cell" style="flex:100 0 auto;min-width:100px;width:100px"><div class="rt-td-inner">Acura</div></div><div class="rt-td rt-align-left" role="cell" style="flex:100 0 auto;min-width:100px;width:100px"><div class="rt-td-inner">Legend__</div></div></div></div><div class="rt-no-data" aria-live="assertive"></div></div></div></div>"`;

exports[`renderToHTML renders tables to HTML with theme styles 1`] = `"<div class="Reactable ReactTable reactable-1galy0v"><div class="rt-table" role="table"><div class="rt-thead" role="rowgroup" style="min-width:200px"><div class="rt-tr rt-tr-header" role="row" style="flex:1 0 auto;min-width:200px"><div class="rt-th rt-align-left reactable-1xdhyk6" role="columnheader" style="position:relative;flex:100 0 auto;min-width:100px;width:100px" aria-label="Sort Manufacturer" aria-sort="none" tabindex="0" data-sort-hint="ascending"><div class="rt-th-inner reactable-1rsvkai"><div class="rt-sort-header"><div class="rt-text-content">Manufacturer</div><span class="rt-sort-right" aria-hidden="true"></span></div></div></div><div class="rt-th rt-align-left reactable-1xdhyk6" role="columnheader" style="position:relative;flex:100 0 auto;min-width:100px;width:100px" aria-label="Sort Model" aria-sort="none" tabindex="0" data-sort-hint="ascending"><div class="rt-th-inner reactable-1rsvkai"><div class="rt-sort-header"><div class="rt-text-content">Model</div><span class="rt-sort-right" aria-hidden="true"></span></div></div></div></div></div><div class="rt-tbody" role="rowgroup" style="min-width:200px"><div class="rt-tr-group"><div class="rt-tr" role="row" style="flex:1 0 auto;min-width:200px"><div class="rt-td rt-align-left reactable-1xdhyk6" role="cell" style="flex:100 0 auto;min-width:100px;width:100px"><div class="rt-td-inner reactable-1rsvkai">Acura</div></div><div class="rt-td rt-align-left reactable-1xdhyk6" role="cell" style="flex:100 0 auto;min-width:100px;width:100px"><div class="rt-td-inner reactable-1rsvkai">Integra</div></div></div></div><div class="rt-tr-group"><div class="rt-tr" role="row" style="flex:1 0 auto;min-width:200px"><div class="rt-td rt-align-left reactable-1xdhyk6" role="cell" style="flex:100 0 auto;min-width:100px;width:100px"><div class="rt-td-inner reactable-1rsvkai">Acura</div></div><div class="rt-td rt-align-left reactable-1xdhyk6" role="cell" style="flex:100 0 auto;min-width:100px;width:100px"><div class="rt-td-inner reactable-1rsvkai">Legend</div></div></div></div><div class="rt-no-data" aria-live="assertive"></div></div></div></div>"`;
